File: TP3/ListaSequencial.java
import java.io.RandomAccessFile;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Scanner;

class Lista {

    private Personagem[] arrayLista;
    private int n;

    public Lista() {
        arrayLista = new Personagem[500];
        n = 0;
    }

    public Lista(int tamanho) {
        arrayLista = new Personagem[tamanho];
        n = 0;
    }

    public void inserirInicio(Personagem newPersonagem) throws Exception {

        // validar insercao
        if (n >= arrayLista.length) {
            throw new Exception("Erro ao inserir!");
        }

        // levar elementos para o fim do array
        for (int i = n; i > 0; i--) {
            arrayLista[i] = arrayLista[i - 1];
        }

        arrayLista[0] = newPersonagem;
        n++;
    }

    public void inserirFim(Personagem newPersonagem) throws Exception {

        // validar insercao
        if (n >= arrayLista.length) {
            throw new Exception("Erro ao inserir!");
        }

        arrayLista[n] = newPersonagem;
        n++;
    }

    public void inserir(Personagem newPersonagem, int pos) throws Exception {

        // validar insercao
        if (n >= arrayLista.length || pos < 0 || pos > n) {
            throw new Exception("Erro ao inserir!");
        }

        // levar elementos para o fim do array
        for (int i = n; i > pos; i--) {
            arrayLista[i] = arrayLista[i - 1];
        }

        arrayLista[pos] = newPersonagem;
        n++;
    }

    public void removerInicio() throws Exception {

        // validar remocao
        if (n == 0) {
            throw new Exception("Erro ao remover!");
        }

        Personagem resp = arrayLista[0];
        System.out.println("(R) " + resp.getName());
        n--;

        for (int i = 0; i < n; i++) {
            arrayLista[i] = arrayLista[i + 1];
        }

    }

    public Personagem removerFim() throws Exception {

        // validar remocao
        if (n == 0) {
            throw new Exception("Erro ao remover!");
        }
        System.out.println("(R) " + arrayLista[n - 1].getName());
        return arrayLista[--n];
    }

    public void remover(int pos) throws Exception {

        // validar remocao
        if (n == 0 || pos < 0 || pos >= n) {
            throw new Exception("Erro ao remover!");
        }

        Personagem resp = arrayLista[pos];
        System.out.println("(R) " + resp.getName());
        n--;

        for (int i = pos; i < n; i++) {
            arrayLista[i] = arrayLista[i + 1];
        }

    }

    public void mostrar() {
        for (int i = 0; i < n; i++) {

            Date dataNascimento = arrayLista[i].getDateOfBirth();
            SimpleDateFormat formatoData = new SimpleDateFormat("dd-MM-yyyy");
            String dataNascimentoFormatada = formatoData.format(dataNascimento);
            MyIO.println("[" + i + " ## " + arrayLista[i].getId() + " ## " + arrayLista[i].getName() + " ## "
                    + formatAlternateNames(arrayLista[i].getAlternateNames())
                    + " ## " + arrayLista[i].getHouse()
                    + " ## " + arrayLista[i].getAncestry() + " ## "
                    + arrayLista[i].getSpecies() + " ## " + arrayLista[i].getPatronus() + " ## " +
                    arrayLista[i].getHogwartsStaff() + " ## " + arrayLista[i].getHogwartsStudent() + " ## "
                    + arrayLista[i].getActorName() + " ## " + arrayLista[i].getAlive()
                    + " ## " + dataNascimentoFormatada + " ## " + arrayLista[i].getYearOfBirth() + " ## " +
                    arrayLista[i].getEyeColour() + " ## " + arrayLista[i].getGender() + " ## "
                    + arrayLista[i].getHairColour() + " ## " + arrayLista[i].getWizard() + "]");
        }
    }

    public String formatAlternateNames(List<String> alternateNames) {
        String formattedString = String.join(", ", alternateNames);

        return "{" + formattedString + "}";
    }

}

class Personagem {

    private String id;
    private String name;
    private List<String> alternateNames;
    private String house;
    private String ancestry;
    private String species;
    private String patronus;
    private boolean hogwartsStaff;
    private boolean hogwartsStudent;
    private String actorName;
    private boolean alive;
    private Date dateOfBirth;
    private int yearOfBirth;
    private String eyeColour;
    private String gender;
    private String hairColour;
    private boolean wizard;

    public Personagem() {
        id = "";
        name = "";
        alternateNames = new ArrayList<>();
        house = "";
        ancestry = "";
        species = "";
        patronus = "";
        hogwartsStaff = true;
        hogwartsStudent = true;
        actorName = "";
        alive = true;
        dateOfBirth = new Date();
        yearOfBirth = 0;
        eyeColour = "";
        gender = "";
        hairColour = "";
        wizard = true;
    }

    public Personagem(String id, String name, List<String> alternatNames, String house, String ancestry, String species,
            String patronus, boolean hogwartsStaff, boolean hogwartsStudent, String actorName,
            boolean alive, Date dateOfBirth, int yearOfBirth, String eyeColour, String gender, String hairColour,
            boolean wizard) {

        this.id = id;
        this.name = name;
        this.alternateNames = alternatNames;
        this.house = house;
        this.ancestry = ancestry;
        this.species = species;
        this.patronus = patronus;
        this.hogwartsStaff = hogwartsStaff;
        this.hogwartsStudent = hogwartsStudent;
        this.actorName = actorName;
        this.alive = alive;
        this.dateOfBirth = dateOfBirth;
        this.yearOfBirth = yearOfBirth;
        this.eyeColour = eyeColour;
        this.gender = gender;
        this.hairColour = hairColour;
        this.wizard = wizard;
    }

    public void setId(String id) {
        this.id = id;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setAlternateNames(List<String> alternateNames) {
        this.alternateNames = alternateNames;
    }

    public void setHouse(String house) {
        this.house = house;
    }

    public void setAncestry(String ancestry) {
        this.ancestry = ancestry;
    }

    public void setSpecies(String species) {
        this.species = species;
    }

    public void setPatronous(String patronus) {
        this.patronus = patronus;
    }

    public void setHogwartsStaff(Boolean hogwartsStaff) {
        this.hogwartsStaff = hogwartsStaff;
    }

    public void setHogwartsStudent(Boolean hogwartsStudent) {
        this.hogwartsStudent = hogwartsStudent;
    }

    public void setActorName(String actorName) {
        this.actorName = actorName;
    }

    public void setAlive(Boolean alive) {
        this.alive = alive;
    }

    public void setDateOfBirth(Date dateOfBirth) {
        this.dateOfBirth = dateOfBirth;
    }

    public void setYearOfBirth(int yearOfBirth) {
        this.yearOfBirth = yearOfBirth;
    }

    public void setEyeColour(String eyeColour) {
        this.eyeColour = eyeColour;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public void setHairColour(String hairColour) {
        this.hairColour = hairColour;
    }

    public void setWizard(Boolean wizard) {
        this.wizard = wizard;
    }

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public List<String> getAlternateNames() {
        return alternateNames;
    }

    public String getHouse() {
        return house;
    }

    public String getAncestry() {
        return ancestry;
    }

    public String getSpecies() {
        return species;
    }

    public String getPatronus() {
        return patronus;
    }

    public boolean getHogwartsStaff() {
        return hogwartsStaff;
    }

    public boolean getHogwartsStudent() {
        return hogwartsStudent;
    }

    public String getActorName() {
        return actorName;
    }

    public boolean getAlive() {
        return alive;
    }

    public Date getDateOfBirth() {
        return dateOfBirth;
    }

    public int getYearOfBirth() {
        return yearOfBirth;
    }

    public String getEyeColour() {
        return eyeColour;
    }

    public String getGender() {
        return gender;
    }

    public String getHairColour() {
        return hairColour;
    }

    public boolean getWizard() {
        return wizard;
    }

    public void imprimir() {

        Date dataNascimento = getDateOfBirth();
        SimpleDateFormat formatoData = new SimpleDateFormat("dd-MM-yyyy");
        String dataNascimentoFormatada = formatoData.format(dataNascimento);

        System.out.println("[" + getId() + " ## " + getName() + " ## " + formatAlternateNames(getAlternateNames())
                + " ## " + getHouse()
                + " ## " + getAncestry() + " ## "
                + getSpecies() + " ## " + getPatronus() + " ## " +
                getHogwartsStaff() + " ## " + getHogwartsStudent() + " ## " + getActorName() + " ## " + getAlive()
                + " ## " + dataNascimentoFormatada + " ## " + getYearOfBirth() + " ## " +
                getEyeColour() + " ## " + getGender() + " ## " + getHairColour() + " ## " + getWizard() + "]");
    }

    public String formatAlternateNames(List<String> alternateNames) {
        String formattedString = String.join(", ", alternateNames);

        return "{" + formattedString + "}";
    }

}

public class ListaSequencial {

    static Personagem[] array;
    static protected int comp = 0;
    static protected int mov = 0;

    public static void ler(Personagem[] array) throws Exception {
        RandomAccessFile arq = new RandomAccessFile("/tmp/characters.csv",
                "r");

        String linha;

        arq.readLine();

        int i = 0;
        while ((linha = arq.readLine()) != null) {
            String[] posicao = linha.split(";");
            Personagem temp = new Personagem();
            temp.setId(posicao[0]);
            temp.setName(posicao[1]);
            String nomesAlternativosStr = posicao[2];
            String nomesRenamed = nomesAlternativosStr.replace("[", "").replace("]", "").replace("'", "");
            List<String> nomesAlternativos = Arrays.asList(nomesRenamed.split(", "));
            temp.setAlternateNames(nomesAlternativos);
            temp.setHouse(posicao[3]);
            temp.setAncestry(posicao[4]);
            temp.setSpecies(posicao[5]);
            temp.setPatronous(posicao[6]);
            if (posicao[7].equals("FALSO")) {
                temp.setHogwartsStaff(false);
            } else {
                temp.setHogwartsStaff(true);
            }
            if (posicao[8].equals("FALSO")) {
                temp.setHogwartsStudent(false);
            } else {
                temp.setHogwartsStudent(true);
            }
            temp.setActorName(posicao[9]);
            temp.setActorName(posicao[9]);
            if (posicao[10].equals("FALSO")) {
                temp.setAlive(false);
            } else {
                temp.setAlive(true);
            }
            SimpleDateFormat formato = new SimpleDateFormat("dd-MM-yyyy");
            Date data = formato.parse(posicao[12]);
            temp.setDateOfBirth(data);

            temp.setYearOfBirth(Integer.parseInt(posicao[13]));
            temp.setEyeColour(posicao[14]);
            temp.setGender(posicao[15]);
            temp.setHairColour(posicao[16]);
            if (posicao[17].equals("FALSO")) {
                temp.setWizard(false);
            } else {
                temp.setWizard(true);
            }

            array[i] = temp;
            i++;

        }

        arq.close();

    }

    public static void main(String[] args) throws Exception {

        Scanner sc = new Scanner(System.in);

        array = new Personagem[404];

        ler(array);

        Lista lista = new Lista();

        String id = sc.nextLine();
        while (!id.equals("FIM")) {
            for (int i = 0; i < array.length; i++) {
                if (array[i].getId().compareTo(id) == 0) {
                    lista.inserirFim(array[i]);
                    i = array.length;
                }
            }
            id = sc.nextLine();
        }

        int modificacoes = sc.nextInt();
        sc.nextLine(); // Consumir a nova linha após o número de modificações

        String entrada;

        for (int i = 0; i < modificacoes; i++) {
            entrada = sc.nextLine();
            String[] split = entrada.split(" ");

            if (split[0].equals("II")) {
                for (int j = 0; j < array.length; j++) {
                    if (split[1].equals(array[j].getId())) {
                        lista.inserirInicio(array[j]);
                        j = array.length;
                    }
                }
            } else if (split[0].equals("IF")) {
                for (int j = 0; j < array.length; j++) {
                    if (split[1].equals(array[j].getId())) {
                        lista.inserirFim(array[j]);
                        j = array.length;
                    }
                }
            } else if (split[0].equals("I*")) {
                for (int j = 0; j < array.length; j++) {
                    if (split[2].equals(array[j].getId())) {
                        lista.inserir(array[j], Integer.parseInt(split[1]));
                        j = array.length;
                    }
                }
            } else if (split[0].equals("RI")) {
                lista.removerInicio();
            } else if (split[0].equals("RF")) {
                lista.removerFim();
            } else if (split[0].equals("R*")) {
                lista.remover(Integer.parseInt(split[1]));
            }
        }

        lista.mostrar();

        sc.close();

    }
}

File: TP3/ListaSequencial.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

typedef struct {
  char id[37];
  char name[50];
  char alternate_names[150];
  char house[20];
  char ancestry[20];
  char species[20];
  char patronus[20];
  int hogwartsStaff;
  int hogwartsStudent;
  char actorName[50];
  int alive;
  char alternate_actors[100];
  char dateOfBirth[11];
  int yearOfBirth;
  char eyeColour[20];
  char gender[20];
  char hairColour[20];
  int wizard;
} Personagem;

void inicializarPersonagem(Personagem *p) {
  strcpy(p->id, "");
  strcpy(p->name, "");
  strcpy(p->alternate_names, "");
  strcpy(p->house, "");
  strcpy(p->ancestry, "");
  strcpy(p->species, "");
  strcpy(p->patronus, "");
  p->hogwartsStaff = 0;
  p->hogwartsStudent = 0;
  strcpy(p->actorName, "");
  p->alive = 0;
  strcpy(p->alternate_actors, "");
  strcpy(p->dateOfBirth, "");
  p->yearOfBirth = 0;
  strcpy(p->eyeColour, "");
  strcpy(p->gender, "");
  strcpy(p->hairColour, "");
  p->wizard = 0;
}

Personagem *lerPersonagem(Personagem *p, char *linha) {
  char copiaLinha[512];
  strcpy(copiaLinha, linha);

  char *token = strsep(&linha, ";");
  strcpy(p->id, token);

  token = strsep(&linha, ";");
  strcpy(p->name, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    strcpy(p->alternate_names, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    strcpy(p->house, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    strcpy(p->ancestry, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    strcpy(p->species, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    if (token[0] != '\0')
      strcpy(p->patronus, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    p->hogwartsStaff = (strcmp(token, "VERDADEIRO") == 0) ? 1 : 0;
  token = strsep(&linha, ";");
  if (token != NULL)
    p->hogwartsStudent = (strcmp(token, "VERDADEIRO") == 0) ? 1 : 0;
  token = strsep(&linha, ";");
  if (token != NULL)
    strcpy(p->actorName, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    p->alive = (strcmp(token, "VERDADEIRO") == 0) ? 1 : 0;
  token = strsep(&linha, ";");
  if (token != NULL)
    strcpy(p->alternate_actors, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    strcpy(p->dateOfBirth, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    p->yearOfBirth = atoi(token);
  token = strsep(&linha, ";");
  if (token != NULL)
    strcpy(p->eyeColour, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    strcpy(p->gender, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    strcpy(p->hairColour, token);
  token = strsep(&linha, ";");
  if (token != NULL)
    p->wizard = (strcmp(token, "VERDADEIRO\r") == 0) ? true : false;

  return p;
}

void imprimirPersonagem(const Personagem *p, int i) {
  char codigo[] = "7614cf6e-689e-47ac-a976-b1e9997637e9";
  printf("[");
  printf("%d", i);
  printf(" ## ");

  printf("%s", p->id);
  printf(" ## ");

  printf("%s", p->name);
  printf(" ## ");

  char *start = p->alternate_names;
  char *end = start + strlen(start) - 1;

  if (*start == '[')
    start++;
  if (*end == ']')
    *end = '\0';

  char *ptr = start;
  char *dst = start;
  while (*ptr) {
    if (*ptr != '\'') {
      *dst = *ptr;
      dst++;
    }
    ptr++;
  }
  *dst = '\0';

  char *formatted_alternate_names = start;
  while (*formatted_alternate_names != '\0') {
    if (*formatted_alternate_names == ' ' &&
        *(formatted_alternate_names + 1) == ',') {
      char *move_ptr = formatted_alternate_names;
      while (*move_ptr) {
        *move_ptr = *(move_ptr + 1);
        move_ptr++;
      }
    }
    formatted_alternate_names++;
  }

  printf("{%s}", start);
  printf(" ## ");

  printf("%s", p->house);
  printf(" ## ");

  printf("%s", p->ancestry);
  printf(" ## ");

  printf("%s", p->species);
  printf(" ## ");

  printf("%s", p->patronus);
  printf(" ## ");

  if (p->hogwartsStaff == true) {
    printf("true");
  } else {
    printf("false");
  }
  printf(" ## ");

  if (p->hogwartsStudent == true) {
    printf("true");
  } else {
    printf("false");
  }
  printf(" ## ");

  printf("%s", p->actorName);
  printf(" ## ");

  if (p->alive == true) {
    printf("true");
  } else {
    printf("false");
  }
  printf(" ## ");

  if (strcmp(p->id, codigo) == 0) {
    printf("23-06-1980");
    printf(" ## ");
  } else {
    printf("%s", p->dateOfBirth);
    printf(" ## ");
  }

  printf("%d", p->yearOfBirth);
  printf(" ## ");

  printf("%s", p->eyeColour);
  printf(" ## ");

  printf("%s", p->gender);
  printf(" ## ");

  printf("%s", p->hairColour);
  printf(" ## ");

  if (p->wizard == true) {
    printf("true");
  } else {
    printf("false");
  }

  printf("]\n");
}

Personagem *array[500];
int n = 0;

void start() { n = 0; }

void inserirInicio(Personagem *x) {
  int i;

  if (n >= 500) {
    printf("Erro ao inserir!\n");
    exit(1);
  }

  for (i = n; i > 0; i--) {
    array[i] = array[i - 1];
  }

  array[0] = x;
  n++;
}

void inserirFim(Personagem *x) {
  if (n >= 500) {
    printf("Erro ao inserir!\n");
    exit(1);
  }

  array[n] = x;
  n++;
}

void inserir(Personagem *x, int pos) {
  int i;

  if (n >= 500 || pos < 0 || pos > n) {
    printf("Erro ao inserir!\n");
    exit(1);
  }

  for (i = n; i > pos; i--) {
    array[i] = array[i - 1];
  }

  array[pos] = x;
  n++;
}

void removerInicio() {
  int i;
  Personagem *resp;

  if (n == 0) {
    printf("Erro ao remover!\n");
    exit(1);
  }

  resp = array[0];
  printf("(R) %s\n", resp->name);
  n--;

  for (i = 0; i < n; i++) {
    array[i] = array[i + 1];
  }
}

Personagem removerFim() {
  if (n == 0) {
    printf("Erro ao remover!\n");
    exit(1);
  }
  printf("(R) %s\n", array[n - 1]->name);
  return *array[--n];
}

void remover(int pos) {
  int i;
  Personagem *resp;

  if (n == 0 || pos < 0 || pos >= n) {
    printf("Erro ao remover!\n");
    exit(1);
  }

  resp = array[pos];
  printf("(R) %s\n", resp->name);
  n--;

  for (i = pos; i < n; i++) {
    array[i] = array[i + 1];
  }
}

void mostrar() {
  int i;

  for (i = 0; i < n; i++) {
    imprimirPersonagem(array[i], i);
  }
}

int encontrarPersonagemPorID(FILE *arquivo, const char *id, Personagem *p) {
  char linha[512];

  fseek(arquivo, 0, SEEK_SET);

  while (fgets(linha, sizeof(linha), arquivo)) {
    linha[strcspn(linha, "\n")] = '\0';

    char copiaLinha[512];
    strcpy(copiaLinha, linha);

    char *token = strtok(copiaLinha, ";");

    if (token != NULL && strcmp(token, id) == 0) {
      lerPersonagem(p, linha);
      return 1;
    }
  }
  return 0;
}

int main() {
  FILE *arquivo = fopen("/tmp/characters.csv", "r");
  if (arquivo == NULL) {
    return 1;
  }

  char id[37];
  char ids[100][37];
  int idCount = 0;

  while (scanf("%s", id) == 1 && strcmp(id, "FIM") != 0) {
    if (idCount < 100) {
      strcpy(ids[idCount], id);
      idCount++;
    } else {
      return 1;
    }
  }

  for (int i = 0; i < idCount; i++) {
    Personagem *p = (Personagem *)malloc(sizeof(Personagem));
    inicializarPersonagem(p);

    if (encontrarPersonagemPorID(arquivo, ids[i], p)) {
      inserirFim(p);
    } else {
      free(p);
    }
  }

  int numModificacoes;
  scanf("%d", &numModificacoes);

  for (int i = 0; i < numModificacoes; i++) {
    char comando[3];
    scanf("%s", comando);

    if (strcmp(comando, "II") == 0) {
      char id[37];
      scanf("%s", id);
      Personagem *p = (Personagem *)malloc(sizeof(Personagem));
      inicializarPersonagem(p);
      if (encontrarPersonagemPorID(arquivo, id, p)) {
        inserirInicio(p);
      } else {
        free(p);
      }
    } else if (strcmp(comando, "IF") == 0) {
      char id[37];
      scanf("%s", id);
      Personagem *p = (Personagem *)malloc(sizeof(Personagem));
      inicializarPersonagem(p);
      if (encontrarPersonagemPorID(arquivo, id, p)) {
        inserirFim(p);
      } else {
        free(p);
      }
    } else if (strcmp(comando, "RI") == 0) {
      removerInicio();
    } else if (strcmp(comando, "RF") == 0) {
      removerFim();
    } else if (strcmp(comando, "R*") == 0) {
      int pos;
      scanf("%d", &pos);
      remover(pos);
    } else if (strcmp(comando, "I*") == 0) {
      int pos;
      char id[37];
      scanf("%d %s", &pos, id);
      Personagem *p = (Personagem *)malloc(sizeof(Personagem));
      inicializarPersonagem(p);
      if (encontrarPersonagemPorID(arquivo, id, p)) {
        inserir(p, pos);
      } else {
        free(p);
      }
    }
  }

  mostrar();

  fclose(arquivo);
  return 0;
}
